lib/filter/filter-netmask6.c

```diff
@@ -114,42 +108,123 @@ _eval(FilterExprNode *s, LogMessage **msgs, gint num_msg)
 {
-  FilterNetmask6 *self = g_new0(FilterNetmask6, 1);
-  struct in6_addr packet_addr;
-  gchar address[INET6_ADDRSTRLEN] = "";
-  gchar *slash = strchr(cidr, '/');
+  const char* slash = strchr(cidr, '/');
+  if (slash)
+    {
+      *slash_position = slash - cidr;
+      return TRUE;
+    }
+  return FALSE;
+}
 
-  if (strlen(cidr) >= INET6_ADDRSTRLEN + 5 || !slash)
+static inline gboolean
+_is_valid_address_len(gchar* address, gint len)
+{
+  return (len > 0) && (len <= INET6_ADDRSTRLEN);
+}
+
+static inline  gboolean
+_is_valid_prefix(gint prefix)
+{
+ return prefix >=1 && prefix <= 128;
+}
+
+static gboolean
+_parse_prefix(gchar *cidr, gint *prefix)
+{
+  const gint default_prefix = 128;
+  gint slash_position;
+
+  if (!_get_slash_position(cidr, &slash_position))
     {
-      strcpy(address, cidr);
-      self->prefix = 128;
+      msg_warning("Prefix is empty. Setting to default (128)",
+                  NULL);
+      *prefix = default_prefix;
+      return TRUE;
     }
-  else
+
+  *prefix = strtol(&cidr[slash_position + 1], NULL, 10);
+  return _is_valid_prefix(*prefix);
+}
+
+static void
+_set_address(gchar* cidr, gint len, gchar address[])
+{
+  strncpy(address, cidr, len);
+  address[len] = 0;
+}
+
+static gint
+_get_address_len(gchar *cidr)
+{
+  gint slash_position;
+
+  if (!_get_slash_position(cidr, &slash_position))
     {
-      self->prefix = atol(slash + 1);
-      if (self->prefix > 0 && self->prefix < 129)
-        {
-          strncpy(address, cidr, slash - cidr);
-          address[slash - cidr] = 0;
-        }
+      return strlen(cidr);
     }
+  return slash_position;
+
+}
 
-  self->is_valid = ((strlen(address) > 0) && inet_pton(AF_INET6, address, &packet_addr) == 1);
-  if (self->is_valid)
-    get_network_address((unsigned char *) &packet_addr, self->prefix, &self->address);
+static gboolean
+_parse_address(gchar *cidr, gchar address[])
+{
+  gint len = _get_address_len(cidr);
+
+  if (!_is_valid_address_len(address, len))
+    return FALSE;
+  _set_address(cidr, len, address);
+  return TRUE;
+}
+
+static gboolean
+_parse(gchar *cidr, gchar address[], gint *prefix)
+{
+  return (_parse_address(cidr, address) && _parse_prefix(cidr, prefix));
+}
+
+FilterExprNode *
+filter_netmask6_new(gchar *cidr)
+{
+  FilterNetmask6 *self = g_new0(FilterNetmask6, 1);
+  gchar address[INET6_ADDRSTRLEN + 1] = {0};
+  struct in6_addr packet_addr;
+  gint prefix = 0;
+
+  filter_expr_node_init_instance(&self->super);
+
+  if (strlen(cidr) == 0)
+    goto filter_netmask6_invalid_error;
+
+  if (!_parse(cidr, address, &prefix))
+    goto filter_netmask6_invalid_error;
+
+  self->prefix = prefix;
+
+  if (inet_pton(AF_INET6, address, &packet_addr) == 1)
+    filter_netmask6_calc_network_address((unsigned char *) &packet_addr, self->prefix, &self->address);
   else
-    self->address = in6addr_loopback;
+    goto filter_netmask6_invalid_error;
 
   self->super.eval = _eval;
   return &self->super;
```

Comment:
Possibly it's a stupid question but we are leaking 'self' here... only `FilterExprNode` will be freed, `FilterNetmask6` isn't. Is it a valid point? @deirf